File: src/store/store.py
"""
ExtractionStore: In-memory data structures for document storage and field indexing.
Uses AVL trees for fast lookups and JSON files for persistence.
"""

import json
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional

from src.store.avl_tree import AVLTree
from src.store.models import DocumentRecord


class ExtractionStore:
    """
    Primary storage for extracted document information.
    
    Components:
    - documents: dict mapping doc_id -> DocumentRecord
    - indexes: dict mapping field_name -> AVLTree (keyed by normalized value)
    
    Persistence:
    - store.json: serialized documents
    - index_<field>.json: inorder traversal of AVL tree for each field
    """
    
    def __init__(self):
        self.documents: Dict[str, DocumentRecord] = {}
        self.indexes: Dict[str, AVLTree] = {}
    
    def add_document(self, record: DocumentRecord) -> None:
        """
        Add a document record to the store.
        
        Args:
            record: DocumentRecord to add
        
        Time complexity: O(1)
        """
        self.documents[record.doc_id] = record
    
    def index_extraction(self, field: str, value: str, doc_id: str) -> None:
        """
        Index an extracted field value for fast retrieval.
        
        Args:
            field: Field name (rule name)
            value: Normalized extracted value
            doc_id: Document ID
        
        Time complexity: O(log n) where n is number of unique values for this field
        """
        if field not in self.indexes:
            self.indexes[field] = AVLTree()
        
        self.indexes[field].insert(value, doc_id)
    
    def query_equals(self, field: str, value: str) -> List[DocumentRecord]:
        """
        Find all documents where field equals value.
        
        Args:
            field: Field name to query
            value: Normalized value to match
        
        Returns:
            List of DocumentRecord objects
        
        Time complexity: O(log n + k) where n is unique values, k is number of results
        """
        if field not in self.indexes:
            return []
        
        doc_ids = self.indexes[field].search(value)
        return [self.documents[doc_id] for doc_id in doc_ids if doc_id in self.documents]
    
    def query_range(self, field: str, low: str, high: str) -> List[DocumentRecord]:
        """
        Find all documents where field value is in range [low, high] (lexicographic).
        
        Args:
            field: Field name to query
            low: Lower bound (inclusive)
            high: Upper bound (inclusive)
        
        Returns:
            List of DocumentRecord objects
        
        Time complexity: O(log n + k) where n is unique values, k is number of results
        """
        if field not in self.indexes:
            return []
        
        doc_ids = self.indexes[field].range_query(low, high)
        return [self.documents[doc_id] for doc_id in doc_ids if doc_id in self.documents]
    
    def list_field_values(self, field: str) -> List[str]:
        """
        List all unique normalized values for a field, sorted.
        
        Args:
            field: Field name
        
        Returns:
            Sorted list of unique values
        
        Time complexity: O(n) where n is number of unique values
        """
        if field not in self.indexes:
            return []
        
        items = self.indexes[field].inorder_items()
        return [key for key, _ in items]
    
    def save(self, base_dir: Path) -> None:
        """
        Persist store to JSON files.
        
        Creates:
        - base_dir/store.json: all documents
        - base_dir/index_<field>.json: inorder traversal for each field index
        
        Args:
            base_dir: Directory to save files to
        """
        base_dir.mkdir(parents=True, exist_ok=True)
        
        # Save documents
        store_path = base_dir / "store.json"
        documents_dict = {}
        for doc_id, record in self.documents.items():
            documents_dict[doc_id] = {
                "doc_id": record.doc_id,
                "file_path": record.file_path,
                "file_type": record.file_type,
                "file_hash": record.file_hash,
                "processed_at": record.processed_at,
                "status": record.status,
                "error_message": record.error_message,
                "extracted": record.extracted,
                "extracted_raw": record.extracted_raw,
            }
        
        with open(store_path, "w", encoding="utf-8") as f:
            json.dump(documents_dict, f, indent=2, ensure_ascii=False)
        
        # Save indexes (inorder traversal)
        for field, tree in self.indexes.items():
            index_path = base_dir / f"index_{field}.json"
            items = tree.inorder_items()
            index_data = [{"key": key, "doc_ids": doc_ids} for key, doc_ids in items]
            
            with open(index_path, "w", encoding="utf-8") as f:
                json.dump(index_data, f, indent=2, ensure_ascii=False)
    
    def load(self, base_dir: Path) -> None:
        """
        Load store from JSON files and rebuild AVL trees.
        
        Args:
            base_dir: Directory to load files from
        """
        # Load documents
        store_path = base_dir / "store.json"
        if store_path.exists():
            with open(store_path, "r", encoding="utf-8") as f:
                documents_dict = json.load(f)
            
            for doc_id, data in documents_dict.items():
                self.documents[doc_id] = DocumentRecord(
                    doc_id=data["doc_id"],
                    file_path=data["file_path"],
                    file_type=data["file_type"],
                    file_hash=data["file_hash"],
                    processed_at=data["processed_at"],
                    status=data["status"],
                    error_message=data.get("error_message"),
                    extracted=data.get("extracted", {}),
                    extracted_raw=data.get("extracted_raw", {}),
                )
        
        # Load indexes and rebuild AVL trees
        for index_file in base_dir.glob("index_*.json"):
            field = index_file.stem.replace("index_", "")
            
            with open(index_file, "r", encoding="utf-8") as f:
                index_data = json.load(f)
            
            tree = AVLTree()
            for item in index_data:
                key = item["key"]
                doc_ids = item["doc_ids"]
                for doc_id in doc_ids:
                    tree.insert(key, doc_id)
            
            self.indexes[field] = tree
    
    def find_by_path_and_hash(self, file_path: str, file_hash: str) -> Optional[DocumentRecord]:
        """
        Find a document by file path and hash (for duplicate detection).
        
        Args:
            file_path: File path
            file_hash: File hash
        
        Returns:
            DocumentRecord if found, None otherwise
        
        Time complexity: O(n) where n is number of documents (linear scan)
        """
        for record in self.documents.values():
            if record.file_path == file_path and record.file_hash == file_hash:
                return record
        return None

File: src/store/store.cpp
#include "document_ingestion/store.hpp"
#include <nlohmann/json.hpp>
#include <fstream>
#include <regex>
#include <algorithm>

namespace document_ingestion {

void ExtractionStore::add_document(const DocumentRecord& record) {
  documents_[record.doc_id] = record;
}

void ExtractionStore::index_extraction(const std::string& field, const std::string& value,
                                       const std::string& doc_id) {
  if (indexes_.find(field) == indexes_.end()) {
    indexes_[field] = AVLTree();
  }
  indexes_[field].insert(value, doc_id);
}

std::vector<DocumentRecord> ExtractionStore::query_equals(const std::string& field,
                                                         const std::string& value) const {
  auto it = indexes_.find(field);
  if (it == indexes_.end()) return {};
  auto doc_ids = it->second.search(value);
  std::vector<DocumentRecord> result;
  for (const auto& id : doc_ids) {
    auto doc_it = documents_.find(id);
    if (doc_it != documents_.end()) result.push_back(doc_it->second);
  }
  return result;
}

std::vector<DocumentRecord> ExtractionStore::query_range(const std::string& field,
                                                          const std::string& low,
                                                          const std::string& high) const {
  auto it = indexes_.find(field);
  if (it == indexes_.end()) return {};
  auto doc_ids = it->second.range_query(low, high);
  std::vector<DocumentRecord> result;
  for (const auto& id : doc_ids) {
    auto doc_it = documents_.find(id);
    if (doc_it != documents_.end()) result.push_back(doc_it->second);
  }
  return result;
}

std::vector<std::string> ExtractionStore::list_field_values(const std::string& field) const {
  auto it = indexes_.find(field);
  if (it == indexes_.end()) return {};
  auto items = it->second.inorder_items();
  std::vector<std::string> result;
  for (const auto& [key, _] : items) result.push_back(key);
  return result;
}

void ExtractionStore::save(const std::filesystem::path& base_dir) const {
  std::filesystem::create_directories(base_dir);
  nlohmann::json docs;
  for (const auto& [doc_id, record] : documents_) {
    nlohmann::json rec;
    rec["doc_id"] = record.doc_id;
    rec["file_path"] = record.file_path;
    rec["file_type"] = record.file_type;
    rec["file_hash"] = record.file_hash;
    rec["processed_at"] = record.processed_at;
    rec["status"] = record.status;
    if (record.error_message) rec["error_message"] = *record.error_message;
    rec["extracted"] = record.extracted;
    rec["extracted_raw"] = record.extracted_raw;
    docs[doc_id] = rec;
  }
  std::ofstream f(base_dir / "store.json");
  f << docs.dump(2);
  for (const auto& [field, tree] : indexes_) {
    nlohmann::json index_data;
    for (const auto& [key, doc_ids] : tree.inorder_items()) {
      index_data.push_back({{"key", key}, {"doc_ids", doc_ids}});
    }
    std::ofstream idx(base_dir / ("index_" + field + ".json"));
    idx << index_data.dump(2);
  }
}

void ExtractionStore::load(const std::filesystem::path& base_dir) {
  auto store_path = base_dir / "store.json";
  if (std::filesystem::exists(store_path)) {
    std::ifstream f(store_path);
    auto docs = nlohmann::json::parse(f);
    for (auto it = docs.begin(); it != docs.end(); ++it) {
      DocumentRecord rec;
      rec.doc_id = it.value()["doc_id"];
      rec.file_path = it.value()["file_path"];
      rec.file_type = it.value()["file_type"];
      rec.file_hash = it.value()["file_hash"];
      rec.processed_at = it.value()["processed_at"];
      rec.status = it.value()["status"];
      if (it.value().contains("error_message") && !it.value()["error_message"].is_null())
        rec.error_message = it.value()["error_message"].get<std::string>();
      if (it.value().contains("extracted"))
        for (auto& [k, v] : it.value()["extracted"].items())
          rec.extracted[k] = v.get<std::string>();
      if (it.value().contains("extracted_raw"))
        for (auto& [k, v] : it.value()["extracted_raw"].items())
          rec.extracted_raw[k] = v.get<std::string>();
      documents_[rec.doc_id] = std::move(rec);
    }
  }
  for (const auto& entry : std::filesystem::directory_iterator(base_dir)) {
    if (!entry.is_regular_file()) continue;
    auto name = entry.path().filename().string();
    if (name.size() > 12 && name.compare(0, 6, "index_") == 0 && name.compare(name.size() - 5, 5, ".json") == 0) {
      std::string field = name.substr(6, name.size() - 6 - 5);  // "index_" + field + ".json"
      std::ifstream idx(entry.path());
      if (!idx) continue;
      try {
        auto index_data = nlohmann::json::parse(idx);
        if (!index_data.is_array()) continue;
        AVLTree tree;
        for (const auto& item : index_data) {
          if (!item.contains("key") || !item.contains("doc_ids")) continue;
          std::string key = item["key"].get<std::string>();
          for (const auto& doc_id : item["doc_ids"]) {
            tree.insert(key, doc_id.get<std::string>());
          }
        }
        indexes_[field] = std::move(tree);
      } catch (...) {}
    }
  }
}

std::vector<std::string> ExtractionStore::get_indexed_fields() const {
  std::vector<std::string> result;
  for (const auto& [k, _] : indexes_) result.push_back(k);
  std::sort(result.begin(), result.end());
  return result;
}

std::optional<DocumentRecord> ExtractionStore::find_by_path_and_hash(
    const std::string& file_path, const std::string& file_hash) const {
  for (const auto& [_, rec] : documents_) {
    if (rec.file_path == file_path && rec.file_hash == file_hash)
      return rec;
  }
  return std::nullopt;
}

}  // namespace document_ingestion
